### PR TITLE
UML-93 PR environments

### PR DESCRIPTION
**Pull request by andrewpearce-digital**
Builds environments based on PR number and branch name

### CODE REVIEW

File: terraform_environment/viewer_ecs.tf [hunk 1]
```diff
@@ -69,13 +72,13 @@ resource "aws_ecs_task_definition" "viewer" {
 // Permissions
 
 resource "aws_iam_role" "use_an_lpa" {
-  name               = "viewer"
+  name               = "${terraform.workspace}-viewer"
```
Comment (nsmithuk) [inline terraform_environment/viewer_ecs.tf]: Maybe add `task_role`?
`"${terraform.workspace}-viewer-task-role"`

File: features/modules.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from collections import defaultdict
+
+def get_frontend_url( frontend ):
+  """
+  Return a URL for a Use an LPA frontend on a specific environment.
+  The environment is based on the Terraform Workspace environment variable.
+  """
+  workspace = os.getenv('TF_WORKSPACE', 'development')
+
+  account_namespace_mapping = defaultdict(lambda: "development.", {'production': "", 'preproduction': "preproduction."})
```
Comment (nsmithuk) [inline features/modules.py]: Probably my lack of Python knowledge, but I don't understand this line. Are you able to add a comment explaining?

File: features/modules.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  dns_account_namespace = account_namespace_mapping[workspace]
+
+  if dns_account_namespace == "development.":
+    dns_env_namespace = workspace + "."
+  else:
+    dns_env_namespace = ""
+
+  URL = 'https://{0}{1}.{2}use-an-lpa.opg.service.justice.gov.uk'.format(dns_env_namespace, frontend, dns_account_namespace)
```
Comment (nsmithuk) [inline features/modules.py]: Typically se only ever use lowercase variable names in Python. Capitals implies that it's a constant.

Comment (andrewpearce-digital) [reply features/modules.py]: Thanks, that's an oversight. Corrected

File: features/modules.py (hunk 2 above)
Comment (andrewpearce-digital) [reply features/modules.py]: Comment added

File: terraform_environment/viewer_ecs.tf (hunk 1 above)
Comment (andrewpearce-digital) [reply terraform_environment/viewer_ecs.tf]: thanks, updated

File: terraform_environment/viewer_ecs.tf [hunk 4]
```diff
@@ -19,13 +19,16 @@ resource "aws_ecs_service" "viewer" {
...
+  # depends_on = ["aws_lb.viewer", "aws_acm_certificate.cert", "aws_acm_certificate_validation.cert"]
+  depends_on = ["aws_lb.viewer"]
 }
 
 //----------------------------------
 // The service's Security Groups
 
 resource "aws_security_group" "ecs_service" {
-  name_prefix = "ecs-service"
+  name_prefix = "${terraform.workspace}-ecs-service"
```
[the remark this answers is not in the document]
Comment (andrewpearce-digital) [reply terraform_environment/viewer_ecs.tf]: updated

File: terraform_environment/secrets.tf [hunk 5]
```diff
@@ -1,12 +1,12 @@
-
 resource "aws_secretsmanager_secret" "session_key" {
```
[the remark this answers is not in the document]
Comment (andrewpearce-digital) [reply terraform_environment/secrets.tf]: default tags added

File: terraform_environment/ecs_cluster.tf [hunk 6]
```diff
@@ -16,7 +16,7 @@ data "aws_iam_policy_document" "task_role_assume_policy" {
 }
 
 resource "aws_iam_role" "execution_role" {
-  name               = "execution_role"
+  name               = "${terraform.workspace}_execution_role"
```
[the remark this answers is not in the document]
Comment (andrewpearce-digital) [reply terraform_environment/ecs_cluster.tf]: namespaced, and replaced _ with -

File: terraform_environment/certificate.tf [hunk 7]
```diff
@@ -1,18 +1,17 @@
-resource "aws_route53_record" "cert_validation" {
```
[the remark this answers is not in the document]
Comment (andrewpearce-digital) [reply terraform_environment/certificate.tf]: deleted

**Pull request merged**